### ISSUE TITLE
How could you calculate the derivative respect to binary weights(-1, +1)?

### ISSUE THREAD

**Issue by tongwu92623** [MatthieuCourbariaux/BinaryConnect#7]
Hi,

I am reading you code and try to understand any single piece. For backward, I got confused on calculating the gradient respect to binary weight. What is the definition of gradient to the discrete value? And also, does theano do this automaticly?

Best regards,

Tong

**Comment by MatthieuCourbariaux** [comment]
Hi,

You can compute the gradient of the loss with respect to the binary weights simply by using the chain-rule. E.g. let:
	
	
	y = Wb . x 
	
where y is the output vector, Wb the binary weights matrix, x the input vector  and L the loss function. Then, using the chain rule, you get:

	dL/dWb = dL/dy . dy/dWb = dL/dy . x

The real tricky part is to backpropagate through the weight's binarization, i.e. computing:

	dL/dW = dL/dWb * dWb/dW 
        where Wb = sign(W)
        and thus dWb/dW = mostly 0

We get around this issue by using what we call the straight-through estimator, i.e. we ignore the dWb/dW term, so that dL/dW = dL/dWb. For more details, I invite you to read the [section 2.3 of the paper](https://arxiv.org/pdf/1511.00363.pdf).

**Comment by tongwu92623** [comment]
Thank you for your quick reply and I got your idea about how to calculate gradient to binary weight, I forgot that Wb is just a variable. Thank you. Am I correct that the theano's round function is not differentiable (I tried the theano.tensor.round () function and try to solve the gradient, and it returned me error), so that you wrote your own round function or sign function which could return the gradient, and the STE is included in the round function.
